man/forw.man
.TH FORW %manext1% 2013-03-22 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
forw \- forward nmh messages
.SH SYNOPSIS
.HP 5
.na
.B forw
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-annotate " | " \-noannotate ]
.RB [ \-form
.IR formfile ]
.RB [ \-format " | " \-noformat ]
.RB [ \-filter
.IR filterfile ]
.RB [ \-inplace " | " \-noinplace ]
.RB [ \-mime " | " \-nomime ]
.RB [ \-draftfolder
.IR +folder ]
.RB [ \-draftmessage
.IR msg ]
.RB [ \-nodraftfolder ]
.RB [ \-editor
.IR editor ]
.RB [ \-noedit ]
.RB [ \-width
.IR columns ]
.RB [ \-from
.IR address ]
.RB [ \-to
.IR address ]
.RB [ \-cc
.IR address ]
.RB [ \-fcc
.IR +folder ]
.RB [ \-subject
.IR text ]
.RB [ \-whatnowproc
.IR program ]
.RB [ \-nowhatnowproc ]
.RB [ \-dashstuffing " | " \-nodashstuffing ]
.RB [ \-build ]
.RB [ \-file
.IR msgfile ]
.HP 5
.B forw
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-digest
.IR list ]
.RB [ \-issue
.IR number ]
.RB [ \-volume
.IR number ]
[other\ switches\ for
.BR forw ]
.ad
.SH DESCRIPTION
.B forw
constructs a new message from a forms (components) file, with a body
composed of the message(s) to be forwarded.  An editor is invoked
and, after editing is complete, the user is prompted before the
message is sent.
.PP
The default message template will direct
.B forw
to construct the draft as follows:
.PP
.RS 5
.nf
From: {from switch} or <Local-Mailbox> or <username@hostname>
To: {to switch} or blank
Fcc: {fcc switch} or +outbox
Subject: {subject switch} or "{original subject} (fwd)"
--------
.fi
.RE
.PP
If a file named
.RI \*(lq forwcomps \*(rq
exists in the user's nmh directory,
it will be used instead of this default form.  You may also specify an
alternate forms file with the switch
.B \-form
.IR formfile .
Forms are processed via the
.B nmh
template system; see
.IR mh\-format (5)
for details.  Components from the first forwarded message are available
as standard component escapes in the forms file.
.PP
In addition to the standard
.I mh\-format
escapes, the following
.I component
escapes are also supported:
.PP
.RS 5
.nf
.ta \w'nmh\-subject	'u +\w'Returns	'u
.I "Escape	Returns	Description"
fcc	string	Any folders specified with `\-fcc\ folder'
nmh\-from	string	Addresses specified with `\-from\ address'
nmh\-to	string	Addresses specified with `\-to\ address'
nmh\-cc	string	Addresses specified with `\-cc\ address'
nmh\-subject	string	Any text specified with `\-subject\ text'
.fi
.RE
.PP
By default, the \*(lqTo:\*(rq and \*(lqcc:\*(rq fields are empty.  You may
add addresses to these fields with the
.B \-to
.I address
and
.B \-cc
.I address
switches.  You may give these switches multiple times to add multiple
addresses.
.PP
By default, the \*(lqFrom:\*(rq field contains either the value of the
.B Local\-Mailbox
profile entry, or a system default email address.
This default can be overridden by using the
.B \-from
.I address
switch.  The default mailbox in the \*(lqFcc:\*(rq field is
.IR +outbox .
This can be overridden by the
.B \-fcc
switch.
.PP
Any text given to the
.B \-subject
switch will be placed in the \*(lqSubject:\*(rq field in the draft.
.PP
If the draft already exists,
.B forw
will ask you as to the disposition of the draft.  A reply of
.B quit
will abort
.BR forw ,
leaving the draft intact;
.B replace
will replace the existing draft with a blank skeleton; and
.B list
will display the draft.
.PP
If the
.B \-annotate
switch is given, each message being forwarded will be annotated
with the lines:
.PP
.RS 5
.nf
Forwarded:\ date
Forwarded:\ addrs
.fi
.RE
.PP
where each address list contains as many lines as required.
This annotation will be done only if the message is sent directly from
.BR forw .
If the message is not sent immediately from
.BR forw ,
.RB \*(lq comp
.BR \-use \*(rq
may be used to re-edit and send the constructed message, but the
annotations won't take place.  Normally, annotations are done in place
in order to preserve any links to the message.
You may change this by using the
.B \-noinplace
switch.
.PP
See
.IR comp (1)
for a description of the
.B \-editor
and
.B \-noedit
switches.
.PP
Although
.B forw
uses a forms (components) file to construct the initial draft,
a message filter file is used to format each forwarded message in the
body of the draft.
The filter file for \fIforw\fR should be a standard form file for
.IR mhl (1),
as
.B forw
will invoke
.B mhl
to filter (re-format) the forwarded messages prior to being output to
the body of the draft.
.PP
The switches
.BR \-noformat ,
.BR \-format ,
and
.B \-filter
.I filterfile
specify which message filter file to use.  If
.B \-noformat
is specified (this is the default), then each forwarded
message is output into the draft exactly as it appears, with no
.B mhl
filtering.  If
.B \-format
is specified, then the following default message filter file,
.RI \*(lq mhl.forward \*(rq ,
which should be adequate for most users, is used:
.PP
.RS 5
.nf
%mhl_forward%
.fi
.RE
.PP
If a file named
.RI \*(lq mhl.forward \*(rq
exists in the user's nmh
directory, it will be used instead.  You may specify an
alternate message filter file with the switch
.B \-filter
.IR filterfile .
.PP
Each forwarded message is separated with an encapsulation delimiter.
By default, any dashes in the first column of the forwarded messages
will be prepended with `\-\ ' so that when received, the message is
suitable for bursting by
.IR burst (1).
This follows the guidelines in RFC 934.  You may use the
.B \-nodashstuffing
switch to suppress this form of quoting.
.PP
Users of
.IR prompter (1)
can, by specifying
.BR prompter 's
.B \-prepend
switch in the
.I \&.mh_profile
file, prepend any commentary text before the forwarded messages.
.PP
To use the MIME rules for encapsulation, specify the
.B \-mime
switch.  This directs
.B forw
to generate an
.B mhbuild
composition file.  Note that
.B nmh
will not invoke
.B mhbuild
automatically; you must specifically give the command
.PP
.RS 5
.nf
What now? mime
.fi
.RE
.PP
prior to sending the draft.
.PP
The
.B \-draftfolder
.I +folder
and
.B \-draftmessage
.I msg
switches invoke the
.B nmh
draft folder facility.  This is an advanced (and highly useful) feature.
Consult
.IR mh-draft (5)
for more information.
.PP
The
.B \-editor
.I editor
switch indicates the editor to use for the initial edit.
Upon exiting from the editor,
.B comp
will invoke the
.B whatnow
program.  See
.IR whatnow (1)
for a discussion of
available options.
The invocation of this program can be inhibited by using the
.B \-nowhatnowproc
switch.  (In fact, it is the
.I whatnow
program which starts the initial edit.  Hence,
.B \-nowhatnowproc
will prevent any edit from occurring.)
.PP
The
.B \-build
switch is intended to be used by the Emacs mh-e interface to
.BR nmh .
It implies
.BR \-nowhatnowproc .
It causes a file <mh-dir>/draft
to be created, containing the draft message that would normally be
presented to the user for editing.  No mail is actually sent.
.PP
The
.B \-file
.I msgfile
switch specifies the message to be forwarded as an exact filename
rather than as an
.B nmh
folder and message number.  This switch implies
.BR \-noannotate .
The forwarded message is simply copied verbatim into the draft;
the processing implied by the
.BR \-filter ,
.BR \-mime ,
and
.B \-digest
switches is bypassed, and the usual leading and
trailing 'Forwarded Message' delimiters are not added.
The same caveats apply to this option as to the
.B \-build
switch.
.PP
The
.B \-digest
.IR list ,
.B \-issue
.IR number ,
and
.B \-volume
.I number
switches implement a digest facility for
.BR nmh .
Specifying these switches enables and/or overloads the following escapes:
.PP
.RS 5
.nf
.ta \w'Component  'u +\w'Escape  'u +\w'Returns  'u
.I "Type	Escape	Returns	Description"
component	digest	string	Argument to `\-digest'
function	cur	integer	Argument to `\-volume'
function	msg	integer	Argument to `\-issue'
.fi
.RE
.SH FILES
.B forw
looks for format and filter files in multiple locations: absolute
pathnames are accessed directly, tilde expansion is done on usernames,
and files are searched for in the user's
.I Mail
directory as specified in their profile.  If not found there, the directory
.RI \*(lq %nmhetcdir% \*(rq
is checked.
.PP
.PD 0
.TP 25
%nmhetcdir%/forwcomps
The default message skeleton.
.TP
<mh-dir>/forwcomps
The user's message skeleton.
.TP
%nmhetcdir%/digestcomps
The default message skeleton if
.B \-digest
is given.
.TP
<mh-dir>/digestcomps
The user's
.B \-digest
skeleton.
.TP
^%nmhetcdir%/mhl.forward
The default message filter.
.TP
<mh-dir>/mhl.forward
The user's message filter.
.TP
^$HOME/.mh_profile
The user's profile.
.PD
.SH "PROFILE COMPONENTS"
.PD 0
.TP 25
Path:
To determine the user's nmh directory.
.TP
Current\-Folder:
To find the default current folder.
.TP
Draft\-Folder:
To find the default draft-folder.
.TP
Editor:
To override the default editor.
.TP
Msg\-Protect:
To set mode when creating a new message (draft).
.TP
fileproc:
Program to refile the message.
.TP
mhlproc:
Program to filter messages being forwarded.
.TP
whatnowproc:
Program to ask the \*(lqWhat now?\*(rq questions.
.PD
.SH "SEE ALSO"
.IR burst (1),
.IR comp (1),
.IR mh\-format (5),
.IR mhbuild (1),
.IR mhl (1),
.IR prompter (1),
.IR repl (1),
.IR send (1),
.IR whatnow (1)
.PP
.I

NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.

uip/forw.c
/* forw.c -- forward a message, or group of messages.
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "whatnowproc.h"
#include "sbr/m_name.h"
#include "sbr/m_gmprot.h"
#include "sbr/getarguments.h"
#include "sbr/read_switch_multiword.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_setcur.h"
#include "sbr/seq_save.h"
#include "sbr/showfile.h"
#include "sbr/smatch.h"
#include "sbr/refile.h"
#include "sbr/cpydata.h"
#include "sbr/cpydgst.h"
#include "sbr/m_draft.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/context_save.h"
#include "sbr/context_replace.h"
#include "sbr/context_find.h"
#include "sbr/ambigsw.h"
#include "sbr/pidstatus.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/arglist.h"
#include "sbr/error.h"
#include <fcntl.h>
#include "h/tws.h"
#include "h/done.h"
#include "h/utils.h"
#include "sbr/m_maildir.h"
#include "forwsbr.h"


#define	IFORMAT	"digest-issue-%s"
#define	VFORMAT	"digest-volume-%s"

#define FORW_SWITCHES \
    X("annotate", 0, ANNOSW) \
    X("noannotate", 0, NANNOSW) \
    X("draftfolder +folder", 0, DFOLDSW) \
    X("draftmessage msg", 0, DMSGSW) \
    X("nodraftfolder", 0, NDFLDSW) \
    X("editor editor", 0, EDITRSW) \
    X("noedit", 0, NEDITSW) \
    X("filter filterfile", 0, FILTSW) \
    X("form formfile", 0, FORMSW) \
    X("format", 5, FRMTSW) \
    X("noformat", 7, NFRMTSW) \
    X("inplace", 0, INPLSW) \
    X("noinplace", 0, NINPLSW) \
    X("mime", 0, MIMESW) \
    X("nomime", 0, NMIMESW) \
    X("digest list", 0, DGSTSW) \
    X("issue number", 0, ISSUESW) \
    X("volume number", 0, VOLUMSW) \
    X("whatnowproc program", 0, WHATSW) \
    X("nowhatnowproc", 0, NWHATSW) \
    X("dashstuffing", 0, BITSTUFFSW)          /* interface to mhl */ \
    X("nodashstuffing", 0, NBITSTUFFSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("file file", 4, FILESW) \
    X("build", 5, BILDSW)                     /* interface from mhe */ \
    X("from address", 0, FROMSW) \
    X("to address", 0, TOSW) \
    X("cc address", 0, CCSW) \
    X("subject text", 0, SUBJECTSW) \
    X("fcc mailbox", 0, FCCSW) \
    X("width columns", 0, WIDTHSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(FORW);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(FORW, switches);
#undef X

#define DISPO_SWITCHES \
    X("quit", 0, NOSW) \
    X("replace", 0, YESW) \
    X("list", 0, LISTDSW) \
    X("refile +folder", 0, REFILSW) \
    X("new", 0, NEWSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(DISPO);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(DISPO, aqrnl);
#undef X

static struct swit aqrl[] = {
    { "quit", 0, NOSW },
    { "replace", 0, YESW },
    { "list", 0, LISTDSW },
    { "refile +folder", 0, REFILSW },
    { NULL, 0, 0 }
};

static char drft[BUFSIZ];

static char delim3[] =
    "\n------------------------------------------------------------\n\n";
static char delim4[] = "\n------------------------------\n\n";


static struct msgs *mp = NULL;		/* used a lot */


/*
 * static prototypes
 */
static void mhl_draft (int, char *, int, int, char *, char *, int);
static void copy_draft (int, char *, char *, int, int, int);
static void copy_mime_draft (int);


int
main (int argc, char **argv)
{
    bool anot = false;
    bool inplace = true;
    bool mime = false;
    int issue = 0, volume = 0, dashstuff = 0;
    bool nedit = false;
    bool nwhat = false;
    int i, in;
    int out, isdf = 0, msgnum = 0;
    int outputlinelen = OUTPUTLINELEN;
    int dat[5];
    char *cp, *cwd, *maildir, *dfolder = NULL;
    char *dmsg = NULL, *digest = NULL, *ed = NULL;
    char *file = NULL, *filter = NULL, *folder = NULL, *fwdmsg = NULL;
    char *from = NULL, *to = NULL, *cc = NULL, *subject = NULL, *fcc = NULL;
    char *form = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    struct stat st;
    struct msgs_array msgs = { 0, 0, NULL };
    bool buildsw = false;

    if (nmh_init(argv[0], true, true)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case ANNOSW: 
		    anot = true;
		    continue;
		case NANNOSW: 
		    anot = false;
		    continue;

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			die("missing argument to %s", argp[-2]);
		    nedit = false;
		    continue;
		case NEDITSW:
		    nedit = true;
		    continue;

		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			die("missing argument to %s", argp[-2]);
		    nwhat = false;
		    continue;
		case BILDSW:
		    buildsw = true;
		    /* FALLTHRU */
		case NWHATSW: 
		    nwhat = true;
		    continue;

		case FILESW: 
		    if (file)
			die("only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;
		case FILTSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    filter = mh_xstrdup(etcpath(cp));
		    mime = false;
		    continue;
		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case FRMTSW:
		    filter = mh_xstrdup(etcpath(mhlforward));
		    continue;
		case NFRMTSW:
		    filter = NULL;
		    continue;

		case INPLSW: 
		    inplace = true;
		    continue;
		case NINPLSW: 
		    inplace = false;
		    continue;

		case MIMESW:
		    mime = true;
		    filter = NULL;
		    continue;
		case NMIMESW: 
		    mime = false;
		    continue;

		case DGSTSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    digest = mh_xstrdup(cp);
		    mime = false;
		    continue;
		case ISSUESW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if ((issue = atoi (cp)) < 1)
			die("bad argument %s %s", argp[-2], cp);
		    continue;
		case VOLUMSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if ((volume = atoi (cp)) < 1)
			die("bad argument %s %s", argp[-2], cp);
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			die("only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
				    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW:
		    if (dmsg)
			die("only one draft message at a time!");
		    if (!(dmsg = *argp++) || *dmsg == '-')
			die("missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case BITSTUFFSW: 
		    dashstuff = 1;	/* ternary logic */
		    continue;
		case NBITSTUFFSW: 
		    dashstuff = -1;	/* ternary logic */
		    continue;

		case FROMSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    from = addlist(from, cp);
		    continue;
		case TOSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    to = addlist(to, cp);
		    continue;
		case CCSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    cc = addlist(cc, cp);
		    continue;
		case FCCSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    fcc = addlist(fcc, cp);
		    continue;
		case SUBJECTSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    subject = mh_xstrdup(cp);
		    continue;

		case WIDTHSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if ((outputlinelen = atoi(cp)) < 10)
			die("impossible width %d", outputlinelen);
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else {
	    app_msgarg(&msgs, cp);
	}
    }

    cwd = mh_xstrdup(pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (file && (msgs.size || folder))
	die("can't mix files and folders/msgs");

try_it_again:

    strncpy (drft, buildsw ? m_maildir ("draft")
			  : m_draft (dfolder, NULL, NOUSE, &isdf), sizeof(drft));

    /* Check if a draft already exists */
    if (!buildsw && stat (drft, &st) != NOTOK) {
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = read_switch_multiword ("\nDisposition? ",
						isdf ? aqrnl : aqrl)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrnl : aqrl)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    dmsg = NULL;
		    goto try_it_again;
		case YESW: 
		    break;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    inform("say what?");
		    break;
	    }
	}
    }

    if (file) {
	/*
	 * Forwarding a file.
         */
	anot = false;	/* don't want to annotate a file */
    } else {
	/*
	 * Forwarding a message.
	 */
	if (!msgs.size)
	    app_msgarg(&msgs, "cur");
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder, 1)))
	    die("unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    die("no messages in %s", folder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < msgs.size; msgnum++)
	    if (!m_convert (mp, msgs.msgs[msgnum]))
		done (1);

	seq_setprev (mp);	/* set the previous sequence */

	/*
	 * Find the first message in our set and use it as the input
	 * for the component scanner
	 */

	for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	    if (is_selected (mp, msgnum)) {
		fwdmsg = mh_xstrdup(m_name(msgnum));
		break;
	    }

	if (! fwdmsg)
	    die("Unable to find input message");
    }

    if (filter && access (filter, R_OK) == NOTOK)
	adios (filter, "unable to read");

    /*
     * Open form (component) file.
     */
    if (digest) {
	if (issue == 0) {
	    snprintf (buf, sizeof(buf), IFORMAT, digest);
	    if (volume == 0
		    && (cp = context_find (buf))
		    && ((issue = atoi (cp)) < 0))
		issue = 0;
	    issue++;
	}
	if (volume == 0) {
	    snprintf (buf, sizeof(buf), VFORMAT, digest);
	    if ((cp = context_find (buf)) == NULL || (volume = atoi (cp)) <= 0)
		volume = 1;
	}
	if (!form)
	    form = digestcomps;
    } else {
    	if (!form)
    	    form = forwcomps;
    }

    dat[0] = digest ? issue : msgnum;
    dat[1] = volume;
    dat[2] = 0;
    dat[3] = outputlinelen;
    dat[4] = 0;


    in = build_form (form, digest, dat, from, to, cc, fcc, subject,
    		     file ? file : fwdmsg);

    if ((out = creat (drft, m_gmprot ())) == NOTOK)
	adios (drft, "unable to create");

    /*
     * copy the components into the draft
     */
    cpydata (in, out, form, drft);
    close (in);

    if (file) {
	/* just copy the file into the draft */
	if ((in = open (file, O_RDONLY)) == NOTOK)
	    adios (file, "unable to open");
	cpydata (in, out, file, drft);
	close (in);
	close (out);
    } else {
	/*
	 * If filter file is defined, then format the
	 * messages into the draft using mhlproc.
	 */
	if (filter)
	    mhl_draft (out, digest, volume, issue, drft, filter, dashstuff);
	else if (mime)
	    copy_mime_draft (out);
	else
	    copy_draft (out, digest, drft, volume, issue, dashstuff);
	close (out);

	if (digest) {
	    snprintf (buf, sizeof(buf), IFORMAT, digest);
	    context_replace (buf, mh_xstrdup(m_str(issue)));
	    snprintf (buf, sizeof(buf), VFORMAT, digest);
	    context_replace (buf, mh_xstrdup(m_str(volume)));
	}

	context_replace (pfolder, folder);	/* update current folder   */
	seq_setcur (mp, mp->lowsel);		/* update current message  */
	seq_save (mp);				/* synchronize sequences   */
	context_save ();			/* save the context file   */
    }

    if (nwhat)
	done (0);
    what_now (ed, nedit, NOUSE, drft, NULL, 0, mp,
	anot ? "Forwarded" : NULL, inplace, cwd, 0);
    done (1);
    return 1;
}


/*
 * Filter the messages you are forwarding, into the
 * draft calling the mhlproc, and reading its output
 * from a pipe.
 */

static void
mhl_draft (int out, char *digest, int volume, int issue,
            char *file, char *filter, int dashstuff)
{
    pid_t child_id;
    int msgnum, pd[2];
    char buf1[BUFSIZ];
    char buf2[BUFSIZ];
    char *program;
    struct msgs_array vec = { 0, 0, NULL };

    if (pipe (pd) == NOTOK)
	adios ("pipe", "unable to create");

    argsplit_msgarg(&vec, mhlproc, &program);

    child_id = fork();
    switch (child_id) {
	case NOTOK: 
	    adios ("fork", "unable to");

	case OK: 
	    close (pd[0]);
	    dup2 (pd[1], 1);
	    close (pd[1]);

	    app_msgarg(&vec, "-forwall");
	    app_msgarg(&vec, "-form");
	    app_msgarg(&vec, filter);

	    if (digest) {
		app_msgarg(&vec, "-digest");
		app_msgarg(&vec, digest);
		app_msgarg(&vec, "-issue");
		snprintf (buf1, sizeof(buf1), "%d", issue);
		app_msgarg(&vec, buf1);
		app_msgarg(&vec, "-volume");
		snprintf (buf2, sizeof(buf2), "%d", volume);
		app_msgarg(&vec, buf2);
	    }

	    /*
	     * Are we dashstuffing (quoting) the lines that begin
	     * with `-'.  We use the mhl default (don't add any flag)
	     * unless the user has specified a specific flag.
	     */
	    if (dashstuff > 0)
		app_msgarg(&vec, "-dashstuffing");
	    else if (dashstuff < 0)
		app_msgarg(&vec, "-nodashstuffing");

	    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
		if (is_selected (mp, msgnum))
		    app_msgarg(&vec, mh_xstrdup(m_name (msgnum)));
	    }

	    app_msgarg(&vec, NULL);

	    execvp (program, vec.msgs);
	    fprintf (stderr, "unable to exec ");
	    perror (mhlproc);
	    _exit(1);

	default: 
	    close (pd[1]);
	    cpydata (pd[0], out, vec.msgs[0], file);
	    close (pd[0]);
	    pidXwait(child_id, mhlproc);
	    break;
    }
}


/*
 * Copy the messages into the draft.  The messages are
 * not filtered through the mhlproc.  Do dashstuffing if
 * necessary.
 */

static void
copy_draft (int out, char *digest, char *file, int volume, int issue, int dashstuff)
{
    int fd,i, msgcnt, msgnum;
    int len, buflen;
    char *bp, *msgnam;
    char buffer[BUFSIZ];

    msgcnt = 1;
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected (mp, msgnum)) {
	    if (digest) {
		strncpy (buffer, msgnum == mp->lowsel ? delim3 : delim4,
			sizeof(buffer));
	    } else {
		/* Get buffer ready to go */
		bp = buffer;
		buflen = sizeof(buffer);

		strncpy (bp, "\n-------", buflen);
		len = strlen (bp);
		bp += len;
		buflen -= len;

		if (msgnum == mp->lowsel) {
		    snprintf (bp, buflen, " Forwarded Message%s",
			PLURALS(mp->numsel));
		} else {
		    snprintf (bp, buflen, " Message %d", msgcnt);
		}
		len = strlen (bp);
		bp += len;
		buflen -= len;

		strncpy (bp, "\n\n", buflen);
	    }
	    if (write (out, buffer, strlen (buffer)) < 0) {
		advise (drft, "write");
	    }

	    if ((fd = open (msgnam = m_name (msgnum), O_RDONLY)) == NOTOK) {
		admonish (msgnam, "unable to read message");
		continue;
	    }

	    /*
	     * Copy the message.  Add RFC934 quoting (dashstuffing)
	     * unless given the -nodashstuffing flag.
	     */
	    if (dashstuff >= 0)
		cpydgst (fd, out, msgnam, file);
	    else
		cpydata (fd, out, msgnam, file);

	    close (fd);
	    msgcnt++;
	}
    }

    if (digest) {
	strncpy (buffer, delim4, sizeof(buffer));
    } else {
	snprintf (buffer, sizeof(buffer), "\n------- End of Forwarded Message%s\n",
		PLURALS(mp->numsel));
    }
    if (write (out, buffer, strlen (buffer)) < 0) {
	advise (drft, "write");
    }

    if (digest) {
	snprintf (buffer, sizeof(buffer), "End of %s Digest [Volume %d Issue %d]\n",
		digest, volume, issue);
	i = strlen (buffer);
	for (bp = buffer + i; i > 1; i--)
	    *bp++ = '*';
	*bp++ = '\n';
	*bp = 0;
	if (write (out, buffer, strlen (buffer)) < 0) {
	    advise (drft, "write");
	}
    }
}


/*
 * Create a mhn composition file for forwarding message.
 */

static void
copy_mime_draft (int out)
{
    int msgnum;
    char buffer[BUFSIZ];

    snprintf (buffer, sizeof(buffer), "#forw [forwarded message%s] +%s",
	PLURALS(mp->numsel), mp->foldpath);
    if (write (out, buffer, strlen (buffer)) < 0) {
	advise (drft, "write");
    }
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	if (is_selected (mp, msgnum)) {
	    snprintf (buffer, sizeof(buffer), " %s", m_name (msgnum));
	    if (write (out, buffer, strlen (buffer)) < 0) {
		advise (drft, "write");
	    }
	}
    if (write (out, "\n", 1) < 0) {
	advise (drft, "write newline");
    }
}
